### ISSUE TITLE
RollingWindow in LiveMode

### ISSUE THREAD

**Issue by AlexCatarino** [QuantConnect/Lean#454]
Using rolling window in live mode, we observe the following bug:

- We add a new bar (bold) to the rolling window indicator (columns are bar open time, price, open, high, low and close, respectively):

**06/07/2016 08:25:00 1.10831** 1.10825 1.1084 1.10791 1.**10831**
06/07/2016 08:24:00 1.10834 1.10821 1.10828 1.10802 1.10834
06/07/2016 08:23:00 1.10817 1.10829 1.1083 1.10809 1.10817

- In the following minute, we add another bar (italic). We note that the previously added bar has changed and the price/close is the same of the newly added bar.

_06/07/2016 08:26:00 1.10812_ 1.10832 1.10842 1.10807 _1.10812_
**06/07/2016 08:25:00 1.10812** 1.10825 1.1084 1.10791 **1.10812**
06/07/2016 08:24:00 1.10834 1.10821 1.10828 1.10802 1.10834

- Next, the bar we are following (bold) keep the previous values with the modified price/close. The other bar (italic) also changed. In this case, we can see that the price/close value is lower than the corresponding low (`1.10801 < 1.10807`), meaning that close was changed without going into `BaseData.Update` method that would set bar.Low to 1.10801. 

06/07/2016 08:27:00 1.10801 1.10811 1.10817 1.10799 1.10801
_06/07/2016 08:26:00 1.10801_ 1.10832 1.10842 1.10807 _1.10801_
**06/07/2016 08:25:00 1.10812** 1.10825 1.1084 1.10791 **1.10812**

In backtest mode, we couldn't observe this bug.
In this bug example, close at history[1] is the same of close at history[0], but it is not always the case. When they are different, they are about 1-3 pips apart.

Code:
```
public class BasicTemplateForexAlgorithm : QCAlgorithm
    {
        private RollingWindow<TradeBar> _rollwin;

        public override void Initialize()
        {
            SetStartDate(2014, 05, 01);  //Set Start Date
            SetEndDate(2014, 05, 15);    //Set End Date
            SetCash(100000);             //Set Strategy Cash
            AddForex("EURUSD", Resolution.Minute);

            _rollwin = new RollingWindow<TradeBar>(5);
        }

        public override void OnData(Slice data)
        {
            _rollwin.Add(data["EURUSD"]);
            if (_rollwin.IsReady)
            {
                for (var i = 0; i < _rollwin.Count; i++)
                {
                    Log("\r\n" + _rollwin[i].Time + " " + _rollwin[i].Price + " " +
                        _rollwin[i].Open + " " +
                        _rollwin[i].High + " " +
                        _rollwin[i].Low + " " +
                        _rollwin[i].Close);
                }
            }
        }
    }
```

**Comment by mchandschuh** [comment]
This looks like reference shenanigans. Maybe we're missing a clone
somewhere. I'll dig into it.

**Comment by jaredbroad** [comment]
Had to revert proposed fix here(https://github.com/QuantConnect/Lean/commit/0b8790bc773cc3155d0fa89f9bc59dd02ae6b15d) as it caused OOM errors. Might need a more discrete/focused fix or a faster cloning system.

**Comment by mchandschuh** [comment]
We could move this logic to TradeBarBuilderEnumerator.MoveNext where we
dequeue from the quasi-single item queue. This guy has the highest
probability of being the culprit, IMO, since he's the one building the
trade bars. If it's not this guy, then we'll have to trace it and figure
out who the hell is modifying that state.

**Comment by jaredbroad** [comment]
Fixed by commit ecdfb09
